lib/request_handler/json_api_data_parser.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require 'request_handler/schema_parser'
+require 'request_handler/error'
+module RequestHandler
+  class JsonApiDataParser < SchemaParser
+    def initialize(data:, schema:, schema_options: {}, included_schemas: {})
+      raise MissingArgumentError, "data": 'is missing' if data.nil?
+      super(schema: schema, schema_options: schema_options)
+      @data = data
+      @included_schemas = included_schemas
+    end
+
+    def run
+      body, *included = flattened_data
+      unless included_schemas?
+        raise SchemaValidationError, included: 'must be empty' unless included.empty?
+        return validate_schema(body)
+      end
+
+      validate_schemas(body, included)
+    end
+
+    private
+
+    def flattened_data
+      body = data.fetch('data') do
```

Comment:
this should also be adapted

`resource = document.fetch("data")`

(same in the error message)